### DISCUSSION TITLE
Embedded Unicode fonts do not render correctly.

### DISCUSSION THREAD

**Discussion by manishpatel0404** [airsdk/Adobe-Runtime-Support#1143] [Q&A] [answered]
I wrote a small utility that allows a user to copy text generated with a legacy TrueType font in a TextField and click a button to convert this text to Unicode text and display it in the adjoining TextField (Screenshots attached).

When I embed the font (Microsoft Shruti in this case), the Unicode rendering is incorrect.

![incorrect_rendering](https://user-images.githubusercontent.com/20116584/132943824-07920729-76d4-498b-b7c8-fb5bf90e174c.png)

However, if I don't embed the font and use the system font instead, the Unicode rendering is as it should be.

![correct_rendering](https://user-images.githubusercontent.com/20116584/132943826-32d6f80b-86de-4af0-b3f4-607e4e0129cb.png)

**Also, when I copy the "incorrect" output and paste it in Notepad or any other Unicode-enabled text processor, the text renders correctly.** This lead me to believe that while the Unicode string is well-formed, there might be a problem in the way embedded Unicode fonts are rendered in the TextField.

**The same erroneous behavior persists even when typing text in a TextField with an embedded Unicode font.**

I am using Windows 7 64-bit and AIR SDK 33.1.575

**Workaround:**
Use system fonts when you have to work with Unicode in TextFields.

However, this might not always be an option when using custom fonts, which are not included with the OS by default. Also, it would be a support nightmare to expect users to install the font themselves. Also, on mobile platforms, I am not sure if it is even possible to install fonts manually or otherwise.

Can somebody throw some light on the subject?

Thanks.

**Comment by ajwfrost** [comment, marked as answer]
Just to check here, the text on the right hand side are showing the Unicode strings, and are within normal text fields? Are there single unicode values for the characters that are displaying incorrectly, or is this where some combinations of values should be joined together to represent a new character?

If this is the former i.e. if you can give us a single unicode character and show that it's displaying wrongly, then I think we'd be able to look at it: but the `TextField` class does not support things like ligatures - you should use the `TextBlock` and `TextLine` classes for this instead...
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/text/engine/TextLine.html

This then uses a proper font rendering engine internally and copes with RTL languages, joining of characters like this, etc..

thanks

**Reply by manishpatel0404** [reply]
Hi,

In both images, I have used `TextField`. The one with device font (lower image) renders the text correctly, while the one with embedded font (upper image) does not.

As an example, construct a string containing 2 Unicode characters U+0A9C (Decimal 2716) and U+0AC0 (Decimal 2752) and assign it to 2 TextFields, one with device font and another with embedded font (Microsoft "Shruti" font) and one can see the difference in rendering. The language is Gujarati, which is not a RTL language.

>  the TextField class does not support things like ligatures

Is this true with embedded fonts only? Because it seems to work with device fonts, as seen in the example above.

Thanks.

**Reply by ajwfrost** [reply]
When I use those characters in Word with some spacing between them, they appear as two separate characters, but if I remove the spaces then they combine into a new, separate character - i.e. a single glyph from the font. This sort of logic is not present in the font rendering code within the normal `TextField` hence if you use embedded fonts, it doesn't work. But yes, if you use device fonts then we just pass the whole thing off to the operating system to draw the text (e.g. on Windows this means the GDI calls such as `TextOut()`) so then the behaviour will be platform-specific.

So for this to work in AIR with embedded fonts, it would need to use the FTE text rendering. That's a little more complex to set up but gives better capabilities..

I believe this all stems from the early days of Flash Player when one of the key goes was to reduce the size of SWF files for dial-up internet connections. So with the initial "embedded fonts" capabilities, the font information that goes into the SWF is very much stripped down/optimised, and we just don't have the information available to be able to handle some of the advanced font topics. To have the full set of font tables that are required for combining characters into new glyphs etc, you need to use the `DefineFont4` tag in the SWF - which is basically embedding an OpenType CFF font, including all the mapping/substitution tables. That's where the `embedAsCFF` flag comes in when you embed your text. CFF fonts can only be used with FTE and non-CFF fonts can only be used with TextField.

thanks

**Reply by manishpatel0404** [reply]
Ah! Thank you for the detailed explanation. Now, everything makes sense. I tried the FTE and it works perfectly. I will switch my code to using FTE in my utility.

Thanks.
